### ISSUE TITLE
Methods wrapped in @lru_cache show type errors

### ISSUE THREAD

**Issue by jonasehrlich** [microsoft/pylance-release#2102]
## Environment data

-   Language Server version: v2021.11.2
-   OS and version: CentOS 7
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.4

## Expected behaviour

Methods without or with optional arguments wrapped in a `@functools.lru_cache` decorator should not show type errors when calling them through an instance of the object.

## Actual behaviour

Since one of the last updates, Pylance reports "_Argument missing for parameter "self"_" even when running in _basic_ type checking mode.

![image](https://user-images.githubusercontent.com/22982787/142989273-fd10c561-af32-4e4d-b7b5-dfeaa690d82f.png)

See the example code in the _Code Snippet_ section below

## Logs

<details>
  <summary>Pylance trace output</summary>
  
  ```log

  [Info  - 9:27:34 AM] Pylance language server 2021.11.2 (pyright b27108d5) starting
  [Info  - 9:27:34 AM] Server root directory: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist
  [Info  - 9:27:34 AM] No configuration file found.
  [Info  - 9:27:34 AM] No pyproject.toml file found.
  [Info  - 9:27:34 AM] Setting pythonPath for service "demo": "/tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/bin/python3"
  [Warn  - 9:27:34 AM] stubPath /home/jonase/demo/typings is not a valid directory.
  [Info  - 9:27:34 AM] Assuming Python version 3.7
  [Info  - 9:27:34 AM] Assuming Python platform Linux
  Search paths for /home/jonase/demo
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib
    /home/jonase/demo
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /home/jonase/demo/typings
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stubs/...
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/bundled/stubs
    /tools/cpkg/site-specific/xdr/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/lib-dynload
    /home/jonase/.local/lib/python3.7/site-packages
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Info  - 9:27:35 AM] Searching for source files
  [Info  - 9:27:35 AM] Found 1 source file
  [Info  - 9:27:35 AM] Background analysis(1) root directory: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist
  [Info  - 9:27:35 AM] Background analysis(1) started
  Background analysis message: setConfigOptions
  Background analysis message: setImportResolver
  Background analysis message: ensurePartialStubPackages
  Background analysis message: setTrackedFiles
  Background analysis message: markAllFilesDirty
  Background analysis message: setFileOpened
  [FG] parsing: /home/jonase/demo/lru-cache-self-missing.py (27ms)
  [FG] parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (89ms)
  [FG] binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (50ms)
  [FG] binding: /home/jonase/demo/lru-cache-self-missing.py (0ms)
  Background analysis message: analyze
  [BG(1)] analyzing: /home/jonase/demo/lru-cache-self-missing.py ...
  [BG(1)]   parsing: /home/jonase/demo/lru-cache-self-missing.py (29ms)
  [BG(1)]   parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (99ms)
  [BG(1)]   binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (40ms)
  [BG(1)]   binding: /home/jonase/demo/lru-cache-self-missing.py (1ms)
  [BG(1)]   checking: /home/jonase/demo/lru-cache-self-missing.py ...
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing.pyi [fs read 2ms] (29ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing.pyi (18ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 1ms] (13ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 1ms] (3ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing_extensions.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/abc.pyi [fs read 1ms] (3ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/abc.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/functools.pyi [fs read 2ms] (6ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/functools.pyi (2ms)
  [BG(1)]   checking: /home/jonase/demo/lru-cache-self-missing.py (123ms)
  [BG(1)] analyzing: /home/jonase/demo/lru-cache-self-missing.py (295ms)
  Background analysis message: resumeAnalysis[Info  - 9:27:34 AM] Pylance language server 2021.11.2 (pyright b27108d5) starting
  [Info  - 9:27:34 AM] Server root directory: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist
  [Info  - 9:27:34 AM] No configuration file found.
  [Info  - 9:27:34 AM] No pyproject.toml file found.
  [Info  - 9:27:34 AM] Setting pythonPath for service "demo": "/tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/bin/python3"
  [Warn  - 9:27:34 AM] stubPath /home/jonase/demo/typings is not a valid directory.
  [Info  - 9:27:34 AM] Assuming Python version 3.7
  [Info  - 9:27:34 AM] Assuming Python platform Linux
  Search paths for /home/jonase/demo
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib
    /home/jonase/demo
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /home/jonase/demo/typings
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stubs/...
    /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/bundled/stubs
    /tools/cpkg/site-specific/xdr/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/site-packages
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7
    /tools/cpkg/.packages/x86_64/RHEL7.2/python/3.7.4/lib/python3.7/lib-dynload
    /home/jonase/.local/lib/python3.7/site-packages
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Warn  - 9:27:35 AM] Exception received when installing file system watcher: TypeError [ERR_FEATURE_UNAVAILABLE_ON_PLATFORM]: The feature watch recursively is unavailable on the current platform, which is being used to run Node.js
  [Info  - 9:27:35 AM] Searching for source files
  [Info  - 9:27:35 AM] Found 1 source file
  [Info  - 9:27:35 AM] Background analysis(1) root directory: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist
  [Info  - 9:27:35 AM] Background analysis(1) started
  Background analysis message: setConfigOptions
  Background analysis message: setImportResolver
  Background analysis message: ensurePartialStubPackages
  Background analysis message: setTrackedFiles
  Background analysis message: markAllFilesDirty
  Background analysis message: setFileOpened
  [FG] parsing: /home/jonase/demo/lru-cache-self-missing.py (27ms)
  [FG] parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (89ms)
  [FG] binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (50ms)
  [FG] binding: /home/jonase/demo/lru-cache-self-missing.py (0ms)
  Background analysis message: analyze
  [BG(1)] analyzing: /home/jonase/demo/lru-cache-self-missing.py ...
  [BG(1)]   parsing: /home/jonase/demo/lru-cache-self-missing.py (29ms)
  [BG(1)]   parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 3ms] (99ms)
  [BG(1)]   binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/builtins.pyi (40ms)
  [BG(1)]   binding: /home/jonase/demo/lru-cache-self-missing.py (1ms)
  [BG(1)]   checking: /home/jonase/demo/lru-cache-self-missing.py ...
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing.pyi [fs read 2ms] (29ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing.pyi (18ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 1ms] (13ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 1ms] (3ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/typing_extensions.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/abc.pyi [fs read 1ms] (3ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/abc.pyi (1ms)
  [BG(1)]     parsing: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/functools.pyi [fs read 2ms] (6ms)
  [BG(1)]     binding: /home/jonase/.vscode-server/extensions/ms-python.vscode-pylance-2021.11.2/dist/typeshed-fallback/stdlib/functools.pyi (2ms)
  [BG(1)]   checking: /home/jonase/demo/lru-cache-self-missing.py (123ms)
  [BG(1)] analyzing: /home/jonase/demo/lru-cache-self-missing.py (295ms)
  Background analysis message: resumeAnalysis

```

</details>

## Code Snippet / Additional information

The following code should not show type errors when running with basic type check:

``` python
import functools

class Config:
    @functools.lru_cache(maxsize=None)
    def foo(self):
        return "foo"

if __name__ == "__main__":
    Config().foo()

```

**Comment by erictraut** [comment]
This is due to a recent change to the typeshed stub file for functools. It looks like the typeshed maintainers may revert the change. See https://github.com/python/typeshed/issues/6347 for additional details.

**Comment by erictraut** [comment]
The change that caused this regression has been reverted in typeshed, and I've incorporated the fix. So this problem should be addressed in the next release of pylance.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.12.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2021120-2-december-2021)
